eventuals/os.h

```diff
@@ -62,6 +66,168 @@ inline void SetAffinity(std::thread& thread, const size_t cpu) {
+
+  PCHECK(pthread_getattr_np(pthread_self(), &attr) == 0)
+      << "Failed to initialize the thread attributes"
+         "object referred to by attr via 'pthread_getattr_np(...)' ";
+
+  PCHECK(pthread_attr_getstack(&attr, &stack_addr, &size) == 0)
+      << "Failed to get stack address via 'pthread_attr_getstack(...)'";
+
+  PCHECK(pthread_attr_destroy(&attr) == 0)
+      << "Failed to destroy thread attributes via "
+         "'pthread_attr_destroy(...)'";
+#endif // If Linux.
+
+#ifdef __MACH__
+  pthread_t self = pthread_self();
+
+  // For macOS `pthread_get_stackaddr_np` can return either the base
+  // or the end of the stack.
+  // Check 1274 line on the link below:
+  // https://android.googlesource.com/platform/art/+/master/runtime/thread.cc
+  void* stack_addr = pthread_get_stackaddr_np(self);
+
+  size_t size = pthread_get_stacksize_np(self);
+#endif // If macOS.
+
+#if defined(__x86_64__)
+#ifdef __linux__
+  // Implementation for Windows might be added in future for this architecture.
+  // For x86 stack grows downward (from highest address to lowest).
+  // (check_line_length skip)
+  // https://eli.thegreenplace.net/2011/02/04/where-the-top-of-the-stack-is-on-x86/
+  // So the stack end will be the lowest address.
+  // The function `pthread_attr_getstack` on Linux returns
+  // the lowest stack address. Since the stack grows downward,
+  // from the highest to the lowest address -> the stack
+  // address from `pthread_attr_getstack` will be the end of
+  // the stack.
+  // https://linux.die.net/man/3/pthread_attr_getstack
+  void* end = stack_addr;
+  void* start = static_cast<char*>(stack_addr) + size;
+  [[maybe_unused]] char local_var{};
+#endif // If Linux.
+
+#ifdef __MACH__
+  void* start = nullptr;
+  void* end = nullptr;
+  [[maybe_unused]] char local_var{};
+
+  // If the address of a local variable is greater than the stack
+  // address from `pthread_get_stacksize_np` ->
+  // `pthread_get_stacksize_np` returns the end of the stack since
+  // on x86 architecture the stack grows downward (from highest to
+  // lowest address).
+  if (&local_var > stack_addr) {
+    start = static_cast<char*>(stack_addr) - size;
+    end = stack_addr;
+  } else {
+    start = stack_addr;
+    end = static_cast<char*>(stack_addr) - size;
+  }
+#endif // If macOS.
+#elif defined(i386) || defined(__i386__) || defined(__i386) || defined(_M_IX86)
+// Implementation might be added in future.
+#elif defined(__powerpc) || defined(__powerpc__) || defined(__powerpc64__)
+  || defined(__POWERPC__) || defined(__ppc__) || defined(__PPC__)
+      || defined(_ARCH_PPC)
+// Implementation might be added in future.
+#elif defined(__PPC64__) || defined(__ppc64__) || defined(_ARCH_PPC64)
+// Implementation might be added in future.
+#elif defined(__sparc__) || defined(__sparc)
+// Implementation might be added in future.
+#elif defined(__arm__) || defined(__arm64__) || defined(__aarch64__)
+// Implementation might be added in future.
```

Comment:
```suggestion
#ifdef __linux__
  // Implementation for Windows might be added in future for this architecture.
  // For x86 stack grows downward (from highest address to lowest).
  // (check_line_length skip)
  // https://eli.thegreenplace.net/2011/02/04/where-the-top-of-the-stack-is-on-x86/
  // So the stack end will be the lowest address.
  // The function `pthread_attr_getstack` on Linux returns
  // the lowest stack address. Since the stack grows downward,
  // from the highest to the lowest address -> the stack
  // address from `pthread_attr_getstack` will be the end of
  // the stack.
  // https://linux.die.net/man/3/pthread_attr_getstack
  return StackInfo{
      /* start = */ static_cast<char*>(stack_addr) + size,
      /* end = */ stack_addr,
      /* size = */ Bytes(size)};
#elif defined(__MACH__)
  // For macOS `pthread_get_stackaddr_np` can return either the start
  // or the end of the stack.
  // Check 1274 line on the link below:
  // https://android.googlesource.com/platform/art/+/master/runtime/thread.cc
  //
  // If the address of a local variable is greater than the stack
  // address from `pthread_get_stacksize_np` ->
  // `pthread_get_stacksize_np` returns the end of the stack since
  // on x86 architecture the stack grows downward (from highest to
  // lowest address).
  [[maybe_unused]] char local_var{};
  if (&local_var > stack_addr) {
    return StackInfo{
        /* start = */ static_cast<char*>(stack_addr) - size,
        /* end = */ stack_addr,
        /* size = */ Bytes(size)};
  } else {
    return StackInfo{
        /* start = */ stack_addr;
        /* end = */ static_cast<char*>(stack_addr) - size,
        /* size = */ Bytes(size)};
  }
#endif
```